### PR TITLE
Add xvf_dfu app

### LINKED ISSUE #68: Extend the DFU application to operate using I2C
This issue follows on from xmos/sw_xvf3800#294 and #64 and covers:

- [ ] Extending the DFU application to allow operation using I2C
- [ ] Updating the corresponding design documentation, and
- [ ] Unit testing the extensions in the DFU application.

System testing of this extension occurs under [xmos/sw_xvf3800#100](https://github.com/xmos/sw_xvf3800/issues/100).

### PR DESCRIPTION
**Pull request by lucianomartin**
Closes #68

### CODE REVIEW

File: src/dfu/dfu_main.cpp
Comment (uvvpavel) [inline src/dfu/dfu_main.cpp:319]: These two options have a dot at the end of a help string. I would suggest removing them as none of the help strings are capitalised and end with a dot.

File: src/dfu/dfu_main.cpp
Comment (uvvpavel) [inline src/dfu/dfu_main.cpp:875]: If the user input protocol is not valid, this API will return a string "Will use USB by default" and use I2C

File: src/dfu/dfu_main.cpp
Comment (uvvpavel) [inline src/dfu/dfu_main.cpp:913]: It is a matter of an app design, but this app will require a device connection even if the user wants to do --help. xvf_host executes all non-device related options before it connects to the device so that the user can execute some options without the hardware. I'm not saying you should do the same, just keep it in mind, it might be a bit nicer.

File: src/dfu/dfu_main.cpp
Comment (uvvpavel) [inline src/dfu/dfu_main.cpp:846]: As far as I see, you don't support --list-commands in this application

File: src/dfu/dfu_main.cpp
Comment (uvvpavel) [inline src/dfu/dfu_main.cpp:1]: Reviewing this file was quite hard for me. I would suggest moving all non-main() things into a separate source file

File: test/test_utils.py
Comment (uvvpavel) [inline test/test_utils.py:58]: why comma?

File: test/test_utils.py
Comment (uvvpavel) [inline test/test_utils.py:85]: I'm not sure why this line has been added. In case of a read command, it will override expect_success variable

File: test/test_options.py
Comment (uvvpavel) [inline test/test_options.py:53]: we should already know that changelog exists and have a path to it from lines 43-44

File: src/dfu/dfu_main.cpp
Comment (uvvpavel) [inline src/dfu/dfu_main.cpp:1]: Error messages in this file have a mixed style (some have been copied from the main app, and some have been written from scratch). In xvf_host error messages do not have "Error:" at the beginning as the message is already piped into stderr and returned -1. Since you're using some utils from the main app, I would suggest that you change your error messaging style so it would be consistent within the dfu app and between both apps.

**Review by uvvpavel — CHANGES_REQUESTED** [review]
Good PR! Few notes and a couple of things to change

File: test/test_utils.py
Comment (lucianomartin) [reply test/test_utils.py:85]: This is needed because the call to `run_cmd()` just below always expects "expect_success = True" if `cmd_vals` is set. I found it quite an unusual implementation choice, and I didn't want to change too much of the original code, so I chose the simplest fix.

File: test/test_options.py
Comment (lucianomartin) [reply test/test_options.py:53]: That is a test for a different host app, so I preferred to keep the check.

File: src/dfu/dfu_main.cpp
Comment (lucianomartin) [reply src/dfu/dfu_main.cpp:1]: I find it helpful to add the "Error:" string, as it is easier to search for errors when standard and error outputs are mixed. I wonder how the warning printouts are written, too.

Anyway, maintaining a similar style to the control host app is more critical. I will remove the  "Error:" string.

File: src/dfu/dfu_main.cpp
Comment (lucianomartin) [reply src/dfu/dfu_main.cpp]: Good catch!

File: src/dfu/dfu_main.cpp
Comment (lucianomartin) [reply src/dfu/dfu_main.cpp:892]: I updated the printout. Now it says for example:

`Will use default driver: device_usb`

File: src/dfu/dfu_main.cpp
Comment (lucianomartin) [reply src/dfu/dfu_main.cpp]: This should be fixed now,

File: src/dfu/dfu_main.cpp
Comment (lucianomartin) [reply src/dfu/dfu_main.cpp]: Good spot. I removed that line.

File: src/dfu/dfu_main.cpp
Comment (lucianomartin) [reply src/dfu/dfu_main.cpp:1]: I created issue #121 to implement the suggest changes after this PR is merged.

**Review by uvvpavel — APPROVED** [review]
All looks good!

**Review by ACascarino — COMMENTED** [review]
I don't speak C++ so my ability to technically critique is limited sadly - picked up a typo
(I also don't have write permission to this repo so my approval doesn't mean much!)

**Pull request merged**
